### PR TITLE
Websocket Test Utilities

### PR DESCRIPTION
**Pull request by ManosMagnus**
Add Utilities in order to establish websocket server with NanoHTTPD. Added the appropriate methods for sending messages and pings.

There are some available tests in my branch [check_test_handshake_send]( https://github.com/ManosMagnus/zap-extensions/tree/check_test_handshake_send)

### CODE REVIEW

File: test/org/zaproxy/zap/testutils/WebSocketTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    private NanoWebSocketTestServer webSocketTestServer;
+    
+    /**
+     * Starts a WebSocket Server on specific hostname and port.
+     */
+    public void startWebSocketServer(String hostname, int webSocketPort) throws IOException {
+        if(webSocketTestServer == null){
+            webSocketTestServer = new NanoWebSocketTestServer(hostname,webSocketPort);
+        }
+        webSocketTestServer.start();
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/WebSocketTestUtils.java]: We should allow to configure the timeout, not all tests might want to use ping/pong to keep the connection alive (I'd expect most to not use them, less "noise").

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Represent a WebSocket Test connection.
+ */
+public class NanoWebSocketConnection extends NanoWSD.WebSocket {
+    private byte[] pingPayload;
+    private Scheduling scheduling;
+    private int closeInterval = 10000;
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java]: This is not used, should be removed?

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+     * Creates a schedule in order to send messages.
+     * @param messages Stack of the messages is going to be sent
+     * @param messageInterval the interval of the messages
+     */
+    public void setOutgoingMessageSchedule(Stack<NanoWSD.WebSocketFrame> messages, int messageInterval){
+        outgoingMessages = messages;
+        scheduling.messageScheduling(messageInterval);
+    }
+    
+    class SendPing extends TimerTask {
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java]: The `TimerTask` is expected to be used with a `Timer`, cancelling this with the thread pool used has no effect (the executor needs to be shut down or the future cancelled).

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                }
+            }
+        }
+    }
+    
+    class Scheduling{
+        ScheduledExecutorService executor;
+        
+        void pingScheduling(int pingInterval){
+            executor = Executors.newScheduledThreadPool(1);
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java]: Better to reuse the executor.

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+/**
+ * Creates a simple WebSocket Test Server. Stores every connection was established
+ */
+public class NanoWebSocketTestServer extends NanoWSD {
+   
+    private List<NanoWebSocketConnection> connectionsList;
+    private final static String DEFAULT_PING_MESSAGE = "PING";
+    
+    public NanoWebSocketTestServer(int port) {
+        super(port);
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: Could call `this(null, port);` instead, which would avoid initialise the `connectionsList` in this constructor.

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    
+    public NanoWebSocketTestServer(String hostname, int port) {
+        super(hostname, port);
+        connectionsList = new ArrayList<>();
+    }
+    
+    /**
+     * Returns the last connection established on server
+     */
+    public NanoWebSocketConnection getTheLastConnection(){
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: Can be just `getLastConnection`. Should provide other methods to get other connections?

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        NanoWebSocketConnection webSocketConnection = new NanoWebSocketConnection(ihttpSession);
+        webSocketConnection.setPingMessage( DEFAULT_PING_MESSAGE.getBytes(StandardCharsets.UTF_8));
+        
+        connectionsList.add(webSocketConnection);
+    
+        return webSocketConnection;
+    }
+    
+    /**
+     * Send the same massage to connected websockets. If connection was closed, removes it from the list
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: Typo in message.

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    public void sendToAll(String str) throws IOException {
+        for (NanoWebSocketConnection connection : connectionsList) {
+            if (connection != null) {
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: Needed? I'd not expect the connections in the list to be null.

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                try {
+                    connection.send(str);
+                } catch (IOException e) {
+                    try {
+                        connection.close(WebSocketFrame.CloseCode.InvalidFramePayloadData,"reqrement",false);
+                    } catch (IOException e1) {
+                        connectionsList.remove(connection);
```
Comment (thc202) [inline test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: Should be done after the catch block? To remove in all cases. (same comments for `disconnectAll`)

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java (hunk 3 above)
Comment (ManosMagnus) [reply test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketConnection.java]: I introduced two Timer one for messages and one for pings. In addition, I introduced a boolean field which checks if the client response on ping message. If the client doesn't respond until the next scheduled ping, the connection will be closed.

File: test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java (hunk 6 above)
Comment (ManosMagnus) [reply test/org/zaproxy/zap/testutils/websocket/server/NanoWebSocketTestServer.java]: I provide also another method in order to get a connection by Channel ID.

**Comment by ManosMagnus** [comment]
Suggestion Addressed

**Comment by thc202** [comment]
Thanks!

**Pull request merged**